src/mcpdotnet/Utils/Json/JsonSerializerOptionsExtensions.cs

```diff
@@ -1,32 +1,67 @@
     /// Creates default options to use for MCP-related serialization.
     /// </summary>
     /// <returns>The configured options.</returns>
+    [UnconditionalSuppressMessage("AotAnalysis", "IL3050", Justification = "DefaultJsonTypeInfoResolver is only used when reflection-based serialization is enabled")]
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "DefaultJsonTypeInfoResolver is only used when reflection-based serialization is enabled")]
     private static JsonSerializerOptions CreateDefaultOptions()
     {
-        JsonSerializerOptions options = new();
+        // If reflection-based serialization is enabled by default, use it, as it's the most permissive in terms of what it can serialize,
+        // and we want to be flexible in terms of what can be put into the various collections in the object model.
+        // Otherwise, use the source-generated options to enable trimming and Native AOT.
 
-        // Add custom converters
-        options.Converters.Add(new JsonRpcMessageConverter());
-        options.Converters.Add(new JsonStringEnumConverter(JsonNamingPolicy.CamelCase));
+        if (JsonSerializer.IsReflectionEnabledByDefault)
+        {
+            // Keep in sync with the JsonSourceGenerationOptions attribute on JsonContext below.
+            JsonSerializerOptions options = new(JsonSerializerDefaults.Web)
+            {
+                TypeInfoResolver = new DefaultJsonTypeInfoResolver(),
+                Converters = { new JsonStringEnumConverter() },
+                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+                NumberHandling = JsonNumberHandling.AllowReadingFromString
+            };
 
-        // Configure general options
-        options.PropertyNameCaseInsensitive = true;
-        options.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull;
-        options.NumberHandling = JsonNumberHandling.AllowReadingFromString;
+            options.MakeReadOnly();
+            return options;
+        }
 
-        return options;
+        return JsonContext.Default.Options;
     }
+
+    internal static JsonTypeInfo<T> GetTypeInfo<T>(this JsonSerializerOptions options) =>
+        (JsonTypeInfo<T>)options.GetTypeInfo(typeof(T));
+
+    // Keep in sync with CreateDefaultOptions above.
+    [JsonSourceGenerationOptions(JsonSerializerDefaults.Web,
+        UseStringEnumConverter = true,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+        NumberHandling = JsonNumberHandling.AllowReadingFromString)]
+    [JsonSerializable(typeof(JsonDocument))]
+    [JsonSerializable(typeof(JsonElement))]
+    [JsonSerializable(typeof(JsonNode))]
+    [JsonSerializable(typeof(IJsonRpcMessage))]
+    [JsonSerializable(typeof(JsonRpcRequest))]
+    [JsonSerializable(typeof(JsonRpcNotification))]
+    [JsonSerializable(typeof(JsonRpcResponse))]
+    [JsonSerializable(typeof(JsonRpcError))]
+    [JsonSerializable(typeof(ServerCapabilities))]
```

Comment:
ClientCapabilities is never serialized?

How do we know this list is complete?

Reply:
It's not an exhaustive list of  all the DTOs, I just put the ones that all the current `JsonSerializer` top-level calls are exercising today.